Multi ssh

mrwulf opened the pull request:
Relates to issues #611, #101, #102, probably others.
 
Added a toolbar with a textbox to run commands in all open putty sessions. Code is roughly based on excellent work by @ozkanpakdil 

There are bunch of improvements that would greatly improve the user experience (and most are fairly straightforward), but I wanted to get this out there so others can use it too!

Future improvements:
1. Add an option to select/limit connections that receive commands
2. Add an option to always open new connections in their own panel (and hide inner tabs)- this would make docking much simpler and easier to see all sessions at once
3. Be able to move a tab between panels
4. Tool strip button to auto-arrange all panels so they are visible
5. Persist recently used commands (maybe even by panel or connection)
6. Don't require typing to be in the toolbar text box (maybe listen to "selected" putty connection and replicate keystrokes to other connections)

sparerd commented:
Very cool - thanks for submitting @mrwulf!

I will hopefully have some time to take a deeper look at this tonight and offer comments.

mrwulf commented:
Awesome, thanks! I haven't done windows development is a while, so please
let me know if I need to make any changes or you have any suggestions.

On Tue, Jul 18, 2017, 8:39 AM David Sparer <notifications@github.com> wrote:

> Very cool - thanks for submitting @mrwulf <https://github.com/mrwulf>!
>
> I will hopefully have some time to take a deeper look at this tonight and
> offer comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mRemoteNG/mRemoteNG/pull/644#issuecomment-316104577>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACYRMVEsfBcdOkkDB9ePWvteNeIB0_xDks5sPNGmgaJpZM4Oanse>
> .
>

sparerd commented on mRemoteV1/UI/Forms/frmMain.cs:
```diff
@@ -961,6 +977,106 @@ private void SetQuickConnectProtocol(string protocol)
 		}
         #endregion
 
+        #region Multi SSH
```
Rather than adding this as a region, let's make this its own class. That way we can reuse it elsewhere if needed and frmMain wont get any bigger

sparerd commented on mRemoteV1/App/NativeMethods.cs:
```diff
@@ -381,6 +384,11 @@ public static int HIWORD(IntPtr value)
         /// <summary>
         /// 
         /// </summary>
+        public const int WM_CHAR = 0x102;
```
Include a summary from MS about this message type:

"The WM_CHAR message contains the character code of the key that was pressed. "

mrwulf commented:
I implemented those changes and also fixed a bug I found- commands wouldn't have been sent to connections established with quick connect (because I was only looking through the ConnectionTree for open connections. Maybe later I'll add another (transient) root node for connections created through quick connect. This hack to tie in connections that are created works well for now.

sparerd commented:
Looks good now. I'll need to talk with @kmscode quick about which branch to target though. We generally don't push/pr directly into master. We might be able to push this into either a new branch or maybe a patch/smaller deployment.

kmscode commented:
@mrwulf had some quick conversations with @sparerd. We're likely going to adjust some of our release plans slightly to get this merged in. We'll merge as soon as we can. Thanks for the PR!!!!

kmscode commented:
@mrwulf discussed this further with @sparerd 

I think there was a lot of users looking for this functionality and the code looks solid enough that we'd like to adjust our release schedule a bit...

Would you mind resubmitting the PR aginst the [patch9 branch](https://github.com/mRemoteNG/mRemoteNG/tree/Patch9)?

Basically patch9 (which was going to be 1.75.7009) will become the new 1.76 and 1.76 will be pushed back to a TBD version number. Patch9 will be going stable "soon".

Thanks again for the awesome PR!

sparerd commented:
@mrwulf I went through some of the merge conflicts an tried to get things working again. Take a look at the following patch of the proposed changes:


[patch.diff.txt](https://github.com/mRemoteNG/mRemoteNG/files/1456248/patch.diff.txt)

sparerd commented:
Also, just noticed that the PuTTYNG.exe file is different. What version is it?

sparerd commented:
Merged this to `develop` in commit c4fdf07
